### PR TITLE
gh-87092: compiler's CFG construction moved to after codegen stage

### PR DESCRIPTION
**Pull request by iritkatriel**
After we merge this I will make a PR to move codegen to a separate file.

Skipping news - we will have one entry for the entire refactor once it's done.

* Issue: gh-87092

### CODE REVIEW

File: Python/compile.c [hunk 1]
```diff
@@ -810,7 +972,7 @@ dictbytype(PyObject *src, int scope_type, int flag, Py_ssize_t offset)
     return dest;
 }
 
-static void
+static bool
 cfg_builder_check(cfg_builder *g)
```
Comment (markshannon) [inline Python/compile.c:938]: Guard this with `#ifndef NDEBUG`?

**Comment by iritkatriel** [comment]
The test failures are a couple of timeouts plus a couple of errors that are also showing up on other branches.

**Comment by markshannon** [comment]
Do we really want another instruction  struct and another instruction sequence struct?
A basic block, extended block and general instruction sequence, are much the same. The main differences being where jumps are allowed and how they are resolved.
The allowed use and meaning of branches should be clear from the context.
If you want stronger typing in C, just wrap the base instruction sequence:
```C
typedef struct InstructionSequence InstructionSequence;
typedef struct {
    InstructionSequence instructions;
} InstructionList;

typedef struct {
    InstructionSequence instructions;
} BasicBlock;
```

The reason I'd like a single InstructionSequence, is that we can stick an object header in front of it and access it from Python.
Having instruction sequences be Python objects would make introspection and debugging *so* much nicer.

**Comment by iritkatriel** [comment]
We can have an InstructionSequence type which is reused in basic blocks, but basic blocks have a lot more fields that we don't need for codegen, and they don't need a label map because they should have pointers to the target blocks.

**Comment by markshannon** [comment]
> We can have an InstructionSequence type which is reused in basic blocks

That makes sense.

**Comment by iritkatriel** [comment]
> > We can have an InstructionSequence type which is reused in basic blocks
> 
> That makes sense.

I'm not sure. 

The instruction in a basicblock is 

```
struct cfg_instr {
    int i_opcode;
    int i_oparg;
    location i_loc;
    /* The following fields should not be set by the front-end: */
    struct basicblock_ *i_target; /* target block (if jump instruction) */
    struct basicblock_ *i_except; /* target block when exception is raised */
};
```

We can replace the first 3 fields by an instruction struct (this makes sense, I only avoided it to keep the diff smaller).   But we would still need the target pointers, or to create a parallel data structure in the basic block for the pointers (and keep it in sync with the instruction sequence when we insert or delete instructions inside a block). It would be simpler to write a function that translates a CFG to something that can be exposed to python.

**Comment by iritkatriel** [comment]
In any case, I don't think we should refactor the CFG data structures in the same PR. It will be another huge change.

**Comment by iritkatriel** [comment]
I've created a function for resizing the doubling arrays, so this logic is not repeated 3 times now.

To share the instruction type between codegen and cfg, we could do:

```
struct cfg_instr {
    instruction i_instr;
    struct basicblock_ *i_target; /* target block (if jump instruction) */
    struct basicblock_ *i_except; /* target block when exception is raised */
};
```

instead of the current:

```
struct cfg_instr {
     int i_opcode;
     int i_oparg;
     location i_loc;
    /* The following fields should not be set by the front-end: */
    struct basicblock_ *i_target; /* target block (if jump instruction) */
    struct basicblock_ *i_except; /* target block when exception is raised */
};
```

Do you think we should do it in this PR? It would make the diff quit a lot messier.

File: Python/compile.c [hunk 2]
```diff
@@ -431,27 +431,73 @@ typedef struct instr_sequence_ {
...
+            new_alloc = index + default_alloc;
+        }
+        size_t newsize = new_alloc * item_size;
+
+        if (oldsize > (SIZE_MAX >> 1)) {
+            PyErr_NoMemory();
+            return ERROR;
+        }
+
+        if (newsize == 0) {
```
Comment (markshannon) [inline Python/compile.c]: Is this possible?

**Review by markshannon — COMMENTED** [review]
A couple of questions, otherwise looks good.

File: Python/compile.c (hunk 2 above)
Comment (iritkatriel) [reply Python/compile.c]: Probably not. Will make it an assert. Thanks.

**Comment by iritkatriel** [comment]
> A couple of questions, otherwise looks good.

Thank you. I'll merge then.

**Pull request merged**